rizinorg/cutter
# Added option to add new flags from hexdump #2932

**Your checklist for this pull request**
- [ X ] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [ X ] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ X ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)


**Detailed description**

- Problem:
    - Currently, users can add new flags from the disassembly widget using the right-click menu and a keyboard shortcut. However, this functionality is missing in the hexdump widget.
- Proposed solution:
    - To address this issue, we propose the following solution:
        - Add Action for Adding New Flag:
            - Introduce a new action in the right-click menu of the hexdump widget for adding a new flag. This action will provide users with a convenient option to add flags directly from the hexdump view.
        - Synchronize with Disassembly Widget:
            - Ensure that the keyboard shortcut for adding flags from the hexdump widget is synchronized with the shortcut used in the disassembly widget. This consistency will enhance the user experience and streamline flag management across different views.
        - Implementation Details:
            - Add a QAction object (actionFlag) within the HexWidget class to represent the action for adding a new flag.
            - Add a QAction object (actionDeleteFlag) within the HexWidget class to represent the action for deleting a flag.
            - Implement the onActionAddFlagTriggered() slot to handle user-triggered events for adding flags within the HexWidget class.
            - Implement the onActionDeleteFlagTriggered() slot to handle user-triggered events for deleting flags within the HexWidget class.
            - Update the context menu event handling (contextMenuEvent()) in the HexWidget class to include the new action for adding flags.
            - Introduce the parentForDialog() function in the HexWidget class to provide the parent widget for dialogs spawned within the hexdump view.

**Test plan (required)**

Functionality Test:
- Open the application and navigate to the hexdump view.
- Right-click in the hexdump widget to bring up the context menu.
- Select the option to add a new flag. Verify that a dialog box or input field appears with the correct address, allowing you to enter a flag name and size.
- Enter a valid flag name and size, then confirm the action.
- Verify that the new flag is correctly added to the list of flags and is visible in the hexdump widget or disassembly widget.

Keyboard Shortcut Synchronization Test:
- While in the hexdump view, use the keyboard shortcut (key bind N) for adding a new flag (ensure this is the same shortcut used in the disassembly widget).
- Confirm that the action for adding a new flag is triggered, similar to the right-click context menu action.
- Follow the same steps as in the functionality test to add a flag and verify its addition.

Cross-Widget Synchronization Test:

- After adding a new flag in the hexdump widget, switch to the disassembly view.
- Verify that the newly added flag is also reflected in the disassembly widget and any other relevant parts of the application where flags are displayed.

![Screenshot 2024-02-15 at 7 02 39 PM](https://github.com/rizinorg/cutter/assets/90643076/e2ea623b-57c9-4929-bcce-ec374c560625)
![Screenshot 2024-02-15 at 7 03 07 PM](https://github.com/rizinorg/cutter/assets/90643076/ca3aeeb1-c7da-48aa-a6e9-46ca9933fa9f)
![Screenshot 2024-02-15 at 7 03 26 PM](https://github.com/rizinorg/cutter/assets/90643076/5d3afbda-89ce-44a4-932a-b189ebc573fd)
![Screenshot 2024-02-15 at 7 03 40 PM](https://github.com/rizinorg/cutter/assets/90643076/79706cfd-26e9-407c-aa9b-d7278ff67131)
![Screenshot 2024-02-15 at 7 03 53 PM](https://github.com/rizinorg/cutter/assets/90643076/d7e29740-3c89-44ce-bc77-908f0efd5fc2)
![Screenshot 2024-02-15 at 7 04 11 PM](https://github.com/rizinorg/cutter/assets/90643076/22fa6817-35e4-487b-b6fb-9dcaaffc5890)

**Closing issues**
